### PR TITLE
LibSanitizer+CMake+CI: Add non-recoverable UBSan support and use it on CI

### PR DESCRIPTION
**Pull request by BertalanD**
I tried to summarize most of the information in the commit messages, but see the "Making LibWeb less heavy on the linker" thread for more background.

~~I've yet to test this with GCC, waiting for results right now.~~

---

**LibSanitizer: Abort on unrecoverable UBSan errors**

There is no sensible recovery possible from non-void functions that
fail to return a value, or from `__builtin_unreachable()` being reached,
all bets are off once they happened. This is why LLVM's libsanitizer
treats these as always fatal, even if all other UBSan checks are made
recoverable with `-fsanitize-recover`. The codegen already assumes that
these handlers will not return.

**LibSanitizer: Add `_abort` variants of UBSan handlers**

These variants make the choice to terminate the process a compile-time
decision rather than a runtime-configurable setting through
`UBSAN_OPTIONS`. As the compiler can treat any code following these
UBSan handler calls as dead code, it allows for significant
optimizations, such as removing redundant UBSan checks and not needing
to worry about saving volatile registers before calling these.

Generation of this kind of instrumentation can be enabled with
`-fsanitize=undefined -fno-sanitize-recover=undefined`. Build system
support will come in a separate PR.

**CMake: Add `UNDEFINED_BEHAVIOR_IS_FATAL` configure option**

This is mainly intended for use on CI, as UBSan instrumentation results
in a serious load and execution time penalty there. See the previous
commit for more details.

With this enabled, the size of LibWeb, built for x86-64 with Clang 17 as
of https://github.com/SerenityOS/serenity/commit/0b91d36a1e640a5784ce2e17a7261b8fb15b34a6 is reduced as follows:
```
      FILE SIZE        VM SIZE
   --------------  --------------
     +18% +2.99Mi  [ = ]       0    .debug_info
     +14%  +758Ki  [ = ]       0    .debug_addr
    +2.6% +7.92Ki  [ = ]       0    .debug_abbrev
    +129% +2.66Ki  [ = ]       0    [Unmapped]
    -0.2%    -208  -0.2%    -208    .plt
    -0.2%    -312  -0.2%    -312    .rela.plt
    -0.1%    -336  -0.1%    -336    .dynsym
    -0.0%    -647  -0.0%    -513    [8 Others]
    -0.1% -1.14Ki  -0.1% -1.14Ki    .dynstr
   -20.1% -53.5Ki -20.1% -53.5Ki    .eh_frame_hdr
    -7.2% -56.8Ki  [ = ]       0    .debug_str_offsets
    -7.1%  -156Ki  [ = ]       0    .debug_str
   -15.0%  -160Ki  [ = ]       0    .symtab
   -63.6%  -245Ki -63.6%  -245Ki    .relr.dyn
   -25.4%  -357Ki -25.4%  -357Ki    .eh_frame
   -27.7% -1.09Mi  [ = ]       0    .strtab
   -59.3% -10.0Mi  [ = ]       0    .debug_rnglists
   -41.3% -11.0Mi  [ = ]       0    .debug_line
   -70.0% -12.0Mi -70.0% -12.0Mi    .rodata
   -65.2% -15.1Mi -65.2% -15.1Mi    .data
   -53.0% -15.7Mi -53.0% -15.7Mi    .text
   -41.7% -62.1Mi -57.7% -43.4Mi    TOTAL
```
**CI: Force UBSan errors to be fatal for Serenity tests**

The test runner script sets the `halt_on_error=1` `UBSAN_OPTIONS` flag
already, this just makes it a compile-time decision. This should
alleviate some of the slowness of running on-target tests without
hardware acceleration.

### CODE REVIEW

File: Userland/Libraries/LibSanitizer/UBSanitizer.cpp [hunk 1]
```diff
@@ -15,50 +15,71 @@ Atomic<bool> AK::UBSanitizer::g_ubsan_is_deadly;
     warnln(fmt, ##__VA_ARGS__);    \
     dbgln("\x1B[31m" fmt "\x1B[0m", ##__VA_ARGS__);
 
+#define ABORT_ALWAYS()                                            \
+    do {                                                          \
+        WARNLN_AND_DBGLN("UBSAN: This error is not recoverable"); \
+        abort();                                                  \
```
Comment (BertalanD) [inline Userland/Libraries/LibSanitizer/UBSanitizer.cpp:21]: hmm, I thought `abort()` would also cause a backtrace to be created, but apparently not :yakfused:

**Comment by BertalanD** [comment]
Even more crazy results with GCC:
```
    FILE SIZE        VM SIZE    
 --------------  -------------- 
  +1.5%  +479Ki  [ = ]       0    .debug_info
  +0.9% +10.0Ki  [ = ]       0    .debug_abbrev
  +240% +8.19Ki  [ = ]       0    [Unmapped]
  +0.0%     +64  [ = ]       0    .debug_str
  -0.0%     -36  -0.0%     -36    [2 Others]
  -0.1%     -64  -0.1%     -64    .plt
  -0.0%     -72  -0.0%     -72    .dynsym
  -0.0%     -92  -0.0%     -92    .dynstr
  -0.1%     -96  -0.1%     -96    .rela.plt
  [ = ]       0  -0.4%    -432    .bss
  -0.2% -1.59Ki  -0.2% -1.59Ki    .rodata
  -2.3% -5.83Ki  -2.3% -5.83Ki    .eh_frame_hdr
  -0.7% -7.50Ki  [ = ]       0    .debug_aranges
  -1.3% -17.4Ki  [ = ]       0    .symtab
  -1.8% -72.5Ki  [ = ]       0    .strtab
  -7.7% -99.3Ki  -7.7% -99.3Ki    .eh_frame
 -60.9%  -550Ki -60.9%  -550Ki    .relr.dyn
 -24.9% -5.46Mi  [ = ]       0    .debug_rnglists
 -37.5% -11.9Mi  [ = ]       0    .debug_line
 -60.8% -35.0Mi -60.8% -35.0Mi    .data
 -61.7% -35.1Mi -61.7% -35.1Mi    .text
 -38.1% -87.7Mi -58.4% -70.7Mi    TOTAL

```

**Review by kleinesfilmroellchen — APPROVED** [review]

**Comment by ADKaster** [comment]
Assuming this works, this looks good to me, nice work! I saw some chatter in #kernel about issues with Clang boots? But the CI passed.

**Comment by BertalanD** [comment]
The Kernel change is not included in this PR.

**Review by ADKaster — APPROVED** [review]

**Pull request merged**